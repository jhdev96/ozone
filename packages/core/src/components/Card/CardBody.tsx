import React from 'react';
import styles from './index.module.scss';


export default function CardBody({
  children
}: {children?: React.ReactNode}): JSX.Element 
{
  return (
    <div className={styles.body}>
      {children}
    </div>
  )
}

CardBody.displayName = 'CardBody';