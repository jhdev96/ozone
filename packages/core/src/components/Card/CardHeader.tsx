import React from 'react';
import styles from './index.module.scss';


export default function CardHeader({
  children
}: {children?: React.ReactNode}): JSX.Element 
{
  return (
    <div className={styles.header}>
      {children}
    </div>
  )
}

CardHeader.displayName = 'CardHeader';