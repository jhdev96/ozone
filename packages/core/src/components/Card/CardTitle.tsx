import React from 'react';
import styles from './index.module.scss';


export default function CardTitle({
  children = "Card Title"
}: {children?: React.ReactNode}): JSX.Element 
{
  return (
    <span className={styles.title}>
      {children}
    </span>
  )
}

CardTitle.displayName = 'CardTitle';