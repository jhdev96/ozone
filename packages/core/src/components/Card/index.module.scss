@import '../../scss/root';
@import '../../scss/mixins/card/card.scss';


.primary {
  @include card;

  &.light {
    color: var(--black-core);

    .header { background-color: #fff; }
    .body { background-color: var(--blue-1); }
  }
  
  &.dark {
    color: var(--gray-1);

    .header { background-color: var(--black-4); }
    .body { background-color: var(--black-6); }
  }
}

.basic {
  @include card;

  &.light {
    background-color: #fff;

    .header {
      color: var(--black-2);
      border-bottom: 1px solid var(--gray-1);
    }
  }
  
  &.dark {
    color: var(--gray-core);
    background-color: var(--black-4);

    .header {
      color: var(--gray-3);
      border-bottom: 1px solid var(--black-3);
    }
  }
}

/* Card elements */
.title { align-self: center; }

.body {
  border-radius: 0 0 6px 6px;
  font-size: 0.9em;
}

/* Card modifiers */
.padding {
  .body { padding: 15px; }
}

.narrow { width: 50%; }
.wide { width: 75%; }
.full { width: 100%; }