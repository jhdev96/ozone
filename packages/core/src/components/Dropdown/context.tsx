import React from 'react';


interface ContextType {
  /** The name of the selected item. */
  selected?: string;
  /** A function that handles selection. */
  handleSelection?: (name: string) => void;
};

export const DropdownContext = React.createContext<ContextType>({
  selected: '',
  handleSelection: () => {}
});