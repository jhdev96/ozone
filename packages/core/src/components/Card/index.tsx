import React from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import CardBody from './CardBody';
import CardHeader from './CardHeader';
import CardTitle from './CardTitle';
import styles from './index.module.scss';


export interface CardProps {
  /** 
   * The type of card being rendered. 
   */
  variation?: "primary" | "basic";
  /** 
   * Gives the card a light or dark appearance. 
   */
  theme?: ThemeType;
  /** 
   * Adds padding to the body of the card. 
   */
  pad?: boolean;
  /** 
   * Determines how much space a card occupies in its container. 
   */
  width?: "narrow" | "wide" | "full";
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * The content rendered inside the card. 
   */
  children?: React.ReactNode;
}

export default function Card({
  variation   = "primary",
  theme       = "light",
  pad         = true,
  width       = "wide",
  dataTestId,
  children
}: CardProps): JSX.Element
{
  return (
    <div 
      className={classNames(
        styles[variation],
        styles[theme],
        pad && styles.padding,
        styles[width]
      )}
      data-testid={dataTestId}
    >
      {children}
    </div>
  );
}

Card.Header = CardHeader;
Card.Title = CardTitle;
Card.Body = CardBody;

Card.displayName = 'Card';