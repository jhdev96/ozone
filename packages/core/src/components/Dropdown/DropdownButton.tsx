import React, { useContext } from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import Colors from '@jhdev96/ozone-colors';
import { Arrow } from '@jhdev96/ozone-icons';
import { DropdownContext } from './context';
import styles from './index.module.scss';


interface DropdownButtonProps {
  /**
   * The type of dropdown being rendered.
   */
  variation?: "basic" | "filter";
  /**
   * Gives the button a light or dark theme.
   */
  theme?: ThemeType;
  /**
   * A function that is called when the button
   * is clicked.
   */
  onClick?: () => void;
  /**
   * Sets the direction of the arrow icon on
   * the button. This will be passed as a prop
   * to arrow icon.
   */
  arrowDirection?: "up" | "down";
  /**
   * This will be true when the dropdown is
   * visible.
   */
  showDropdown: boolean;
}

export default function DropdownButton({
  variation,
  theme,
  onClick,
  arrowDirection,
  showDropdown
}: DropdownButtonProps): JSX.Element 
{
  const {selected} = useContext(DropdownContext);

  return (
    <div
      onClick={onClick} 
      className={classNames(
        styles[`${variation}__button`], 
        styles[`${variation}__button--${theme}`],
        showDropdown && styles[`${variation}__button--visible`]
      )}
    >
      <span className={styles[`${variation}__title`]}>
        {selected}
      </span>
      <Arrow 
        direction={arrowDirection} 
        color={Colors.grayScale.BLACK2} 
      />
    </div>
  )
}