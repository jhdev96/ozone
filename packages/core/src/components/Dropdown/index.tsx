import React, { useState, useRef, LegacyRef } from 'react';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import { DropdownContext } from './context';
import DropdownButton from './DropdownButton';
import Manager from './Manager';
import Item from './Item';
import styles from './index.module.scss';


interface DropdownProps {
  /** 
   * Sets the theme for the dropdown. 
   */
  theme?: ThemeType;
  /** 
   * Main classnames for dropdowns. 
   */
  variation?: "basic" | "filter";
  /** 
   * When true, this will display 
   * the selected option as the new title. 
   */
  saveSelection?: Boolean;
  /** 
   * Give the dropdown a default option when rendered. 
   */
  defaultValue?: string;
  /** 
   * The function to be invoked when a selection is made. 
   * A `selectedItem` parameter should be given to the
   * callback so that when an option is selected, its
   * value is received by the dropdown's parent element.
   */
  onSelection?: (selectedItem: string) => void;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /**
   * The dropdown menu items.
   */
  children?: React.ReactNode;
}

export default function Dropdown({
  theme           = "light", 
  variation       = "basic",
  saveSelection,
  defaultValue    = "Dropdown",
  onSelection     = () => {},
  dataTestId,
  children
}: DropdownProps): JSX.Element 
{
  const [selected, setSelected] = useState<string>(defaultValue);
  const [visible, setVisible] = useState(false);
  const [aboveTarget, setAboveTarget] = useState(false);
  const wrapperRef = useRef<HTMLDivElement>();
  const menuRef = useRef<HTMLDivElement>();

  const toggleDropdown = () => {
    setVisible(!visible);
    setAboveTarget(false);
  }

  const handleSelection = (name: string) => {
    /** 
     * display the selected option from the dropdown menu 
     * if the `saveSelection` prop is set to true.
     */
    if (saveSelection) setSelected(name); 
    
    name && onSelection(name);
    toggleDropdown();
  }
       
  return (
    <Manager
      clickRef={wrapperRef}
      overflowRef={menuRef}
      className={classNames(
        styles[variation],
        aboveTarget && styles.aboveTarget
      )}
      onClick={() => {visible && toggleDropdown()}}
      setStateFn={setAboveTarget}
      showDropdown={visible}
    >
      <DropdownContext.Provider value={{selected, handleSelection}}>
        <DropdownButton
          variation={variation}
          theme={theme}
          onClick={toggleDropdown}
          showDropdown={visible}
          arrowDirection={aboveTarget ? "up" : "down"}
        />
        <div 
          className={classNames(
            styles['dropdown__menu'],
            styles[`dropdown__menu--${variation}`],
            styles[`dropdown__menu--${variation}--${theme}`],
            visible && styles.visible
          )}
          ref={menuRef as LegacyRef<HTMLDivElement>}
          data-testid={dataTestId}
        >
          {children}
        </div>
      </DropdownContext.Provider>
    </Manager>
  )
}

Dropdown.displayName = 'Dropdown';
Dropdown.Item = Item;