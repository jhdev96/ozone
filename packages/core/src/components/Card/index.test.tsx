import React from 'react';
import { render, screen } from '@testing-library/react';
import Card from '.';
import Dropdown from '../Dropdown';


describe('<Card />', () => {
  it('renders a `primary` card by default', () => {
    const {container} = render(
      <Card>This is a card</Card>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a card with padding', () => {
    const {container} = render(
      <Card pad dataTestId="test"></Card>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders the correct width', () => {
    render(<Card width="narrow" dataTestId="narrow"></Card>);
    render(<Card width="wide" dataTestId="wide"></Card>);
    render(<Card width="full" dataTestId="full"></Card>);

    expect(screen.getByTestId('narrow')).toMatchSnapshot();
    expect(screen.getByTestId('wide')).toMatchSnapshot();
    expect(screen.getByTestId('full')).toMatchSnapshot();
  });

  it('renders a card with a dark theme', () => {
    render(
      <Card theme="dark" dataTestId="test">text</Card>
    );

    expect(screen.getByTestId('test')).toMatchSnapshot();
  });

  it('renders a card with a dropdown', () => {
    render(
      <Card 
        variation="basic"
        dataTestId="card"
      >
        <Card.Header>
          <Card.Title>Title</Card.Title>
          <Dropdown 
            variation="basic"
            dataTestId="dropdown"
          >
            <Dropdown.Item>A</Dropdown.Item>
            <Dropdown.Item>B</Dropdown.Item>
            <Dropdown.Item>C</Dropdown.Item>
          </Dropdown>
        </Card.Header>
        text
      </Card>
    );

    const card = screen.getByTestId('card');
    const dropdown = screen.getByTestId('dropdown');
    expect(card).toContainElement(dropdown);
  });
});