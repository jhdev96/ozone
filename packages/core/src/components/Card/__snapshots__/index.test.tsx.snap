// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> renders a \`primary\` card by default 1`] = `
<div>
  <div
    class="primary light padding wide"
  >
    This is a card
  </div>
</div>
`;

exports[`<Card /> renders a card with a dark theme 1`] = `
<div
  class="primary dark padding wide"
  data-testid="test"
>
  text
</div>
`;

exports[`<Card /> renders a card with padding 1`] = `
<div>
  <div
    class="primary light padding wide"
    data-testid="test"
  />
</div>
`;

exports[`<Card /> renders the correct width 1`] = `
<div
  class="primary light padding narrow"
  data-testid="narrow"
/>
`;

exports[`<Card /> renders the correct width 2`] = `
<div
  class="primary light padding wide"
  data-testid="wide"
/>
`;

exports[`<Card /> renders the correct width 3`] = `
<div
  class="primary light padding full"
  data-testid="full"
/>
`;
